test/cql-pytest/test_native_transport.py

```diff
@@ -0,0 +1,31 @@
+# Copyright 2023-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+import pytest
+import time
+import nodetool
+from util import new_test_table
+from cassandra.cluster import NoHostAvailable
+
+def test_enable_disable_binary(cql, test_keyspace):
+    schema = 'k text, v text, primary key (k)'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        cql.execute(f"INSERT INTO {table} (k,v) VALUES ('foo', 'bar')")
+        cql.execute(f"SELECT * FROM {table}")
+
+        nodetool.disablebinary(cql)
+        with pytest.raises(NoHostAvailable):
+            cql.execute(f"SELECT * FROM {table}")
+
+        nodetool.enablebinary(cql)
+        pause = 0.1
+        while pause < 100:
+            try:
+                cql.execute(f"SELECT * FROM {table}")
+                break
+            except NoHostAvailable:
+                time.sleep(pause)
+                pause += pause
```

Comment:
I'm curious how long does this take typically. Does it simply wait for Scylla to start everything back up, or does it rely on some Python driver mechanism which needs to time out for something (why?).

Also, to be honest, I still don't understand the thing with test.py. If the connection is back alive (you waited for it here), why doesn't test.py know it? I see it checks the "control connection" but why isn't the "control connection" back on as well, and if it isn't - isn't it a problem for other tests following this one if you run all the tests together in one pytest (i.e., `test/cql-pytest/run`)?  Is it a python driver bug, that if the "control connection" breaks it is never restored? That's very strange.
(by the way, I don't actually know what the "control connection" is, or why test.py chose to check the "control connection" instead of the usual connection that is used for usual CQL commands...).

Reply:
> I'm curious how long does this take typically. Does it simply wait for Scylla to start everything back up, or does it rely on some Python driver mechanism which needs to time out for something (why?).

Didn't check it, but from logs cql session is back to work in 1-2 seconds after the scylla-side listening port is ready. Do you think that us not completely understanding the driver's reconnection logic is blocker for this PR?

> Also, to be honest, I still don't understand the thing with test.py. If the connection is back alive (you waited for it here), why doesn't test.py know it?

Connection is not back alive, it's listening socket that's back alive and it's up to the client to reconnect (close old socket and open new one). The fixture's cql client reconnects "eventually" (as per above) and test does waits for it, but test.py's doesn't wait for its control connection socket to reconnect.

> isn't it a problem for other tests following this one if you run all the tests together in one pytest (i.e., test/cql-pytest/run)? 

Shouldn't be, this test waits for the fixture's cql session to restore, so any other test that re-uses it will continue working normally. No tests use test.py's control connection, so even if it's "broken" only test.py can notice it

> by the way, I don't actually know what the "control connection" is, or why test.py chose to check the "control connection" instead of the usual connection that is used for usual CQL commands..

It's regular cql session object from driver, "control_connection" is just class member name

Reply:
> Didn't check it, but from logs cql session is back to work in 1-2 seconds after the scylla-side listening port is ready. Do you think that us not completely understanding the driver's reconnection logic is blocker for this PR?

I'm not worried about the lack of full understanding, I'm more worried that maybe what we don't understand will break test/cql-pytest/run (i.e., running all the tests in one Scylla and one pytest invocation) and maybe something else doesn't work if you saw that the "control connection" didn't work. I'm worried that adding more code (or configuration options) to test.py just hides the fact there are real problems that can bite us elsewhere.

I was also, to be honest, worried that maybe it isn't 1-2 seconds and maybe it's 10-20 seconds - so if you saw it's 1-2 seconds, it's acceptable.

> > Also, to be honest, I still don't understand the thing with test.py. If the connection is back alive (you waited for it here), why doesn't test.py know it?
> 
> Connection is not back alive, it's listening socket that's back alive and it's up to the client to reconnect (close old socket and open new one). The fixture's cql client reconnects "eventually" (as per above) and test does waits for it, but test.py's doesn't wait for its control connection socket to reconnect.

Right, so maybe if this test if it knows it broke the "connection" and waits for it to recover, maybe if it knows also broke the "control connection" it should also wait for that to recover? I.e., checking the same thing that test.py checks, in a loop?. If this works, and the control connection does recover, it improves two things: 1. you don't need to change test.py, 2. if another test running on the same pytest invocation needs the control connection, it will have it.

> > isn't it a problem for other tests following this one if you run all the tests together in one pytest (i.e., test/cql-pytest/run)?
> 
> Shouldn't be, this test waits for the fixture's cql session to restore, so any other test that re-uses it will continue working normally. No tests use test.py's control connection, so even if it's "broken" only test.py can notice it

I don't know what exactly the "control connection" is (and why test.py chose to check it), but I assume it's something that is needed also in regular operation internally - maybe to detect schema changes or something, I don't know. So I'm not sure what's the implication of continuing to run while it's broken. Did you try running "test/cql-pytest/run" (to run all the tests one after another on the same Scylla and pytest) and see if it still works?

> > by the way, I don't actually know what the "control connection" is, or why test.py chose to check the "control connection" instead of the usual connection that is used for usual CQL commands..
> 
> It's regular cql session object from driver, "control_connection" is just class member name

I think control_connection is a feature of the Python driver which it uses internally for something - it's not an ordinary connection that test.py created for its own use.

Reply:
> > > Also, to be honest, I still don't understand the thing with test.py. If the connection is back alive (you waited for it here), why doesn't test.py know it?
> > 
> > 
> > Connection is not back alive, it's listening socket that's back alive and it's up to the client to reconnect (close old socket and open new one). The fixture's cql client reconnects "eventually" (as per above) and test does waits for it, but test.py's doesn't wait for its control connection socket to reconnect.
> 
> Right, so maybe if this test if it knows it broke the "connection" and waits for it to recover, maybe if it knows also broke the "control connection" it should also wait for that to recover? 
> I.e., checking the same thing that test.py checks, in a loop?. If this works, and the control connection does recover, it improves two things: 1. you don't need to change test.py, 2. if another test running on the same pytest invocation needs the control connection, it will have it.

Agree. It was in v1 of the PR. I dropped this and changed it to mark the cluster dirty and stop re-using it, because this seemed more natural -- the test distracted the cluster in unusual manner so we better not expect it to be re-usable. I don't have strong preference of any of the approaches though.

> > > isn't it a problem for other tests following this one if you run all the tests together in one pytest (i.e., test/cql-pytest/run)?
> > 
> > 
> > Shouldn't be, this test waits for the fixture's cql session to restore, so any other test that re-uses it will continue working normally. No tests use test.py's control connection, so even if it's "broken" only test.py can notice it
> 
> I don't know what exactly the "control connection" is (and why test.py chose to check it), but I assume it's something that is needed also in regular operation internally - maybe to detect schema changes or something, I don't know. So I'm not sure what's the implication of continuing to run while it's broken. 

The control connection is the pylib/scylla_cluster.py's connection that's used for one thing -- at the end of the test it checks that no keyspaces were left by the test, so that new tests don't step on them accidentally.

> Did you try running "test/cql-pytest/run" (to run all the tests one after another on the same Scylla and pytest) and see if it still works?

Yes, it still works. And, for extra confidence, this native-transport-breaking test case is run at 72% mark, so there are plenty of other tests that continue running and passing afterwards

> > > by the way, I don't actually know what the "control connection" is, or why test.py chose to check the "control connection" instead of the usual connection that is used for usual CQL commands..
> > 
> > 
> > It's regular cql session object from driver, "control_connection" is just class member name
> 
> I think control_connection is a feature of the Python driver which it uses internally for something - it's not an ordinary connection that test.py created for its own use.

Quite opposite, as I wrote above. The control connection _is_ the test.py's pylib internal thing that's used to verify that tests re-using the same cluster don't see each other's trash

Reply:
@nyh , so what's your preference here -- should the test.py's control connection be taught to self-heal, or current "throw the whole cluster away" approach is good enough?

Reply:
ping @nyh. please see @xemul's question above.
let's drive this PR to closure please.

Reply:
@nyh , reminder ping

Reply:
@nyh , reminder ping

Reply:
@nyh , please, this PR needs your attention badly